core/iwasm/compilation/aot_emit_memory.c

```diff
@@ -295,24 +329,81 @@ aot_check_memory_overflow(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
+                                               offset, bytes))
+                    goto fail;
+            }
         }
+        else {
+            LLVMValueRef maddr1, maddr2;
+
+            ADD_BASIC_BLOCK(merge_block, "merge_block");
+            ADD_BASIC_BLOCK(runtime_bounds_check, "runtime_bounds_check");
+            ADD_BASIC_BLOCK(check_succ, "check_succ");
+            
+            LLVMBuildCondBr(comp_ctx->builder, cmp, runtime_bounds_check,
+                            check_succ);
+            if (is_local_of_aot_value) {
+                if (!aot_checked_addr_list_add(func_ctx, local_idx_of_aot_value,
+                                               offset, bytes))
+                    goto fail;
+            }
 
-        SET_BUILD_POS(check_succ);
+            SET_BUILD_POS(runtime_bounds_check);
+            maddr1 = aot_call_runtime_bounds_check(comp_ctx, func_ctx, offset1,
+                                                  bytes);
+            LLVMBuildBr(comp_ctx->builder, merge_block);
 
-        if (is_local_of_aot_value) {
-            if (!aot_checked_addr_list_add(func_ctx, local_idx_of_aot_value,
-                                           offset, bytes))
-                goto fail;
+            SET_BUILD_POS(check_succ);
+            if (!enable_segue) {
+                /* maddr = mem_base_addr + offset1 */
+                if (!(maddr2 = LLVMBuildInBoundsGEP2(comp_ctx->builder,
+                                                    INT8_TYPE, mem_base_addr,
+                                                    &offset1, 1, "maddr"))) {
+                    aot_set_last_error("llvm build add failed.");
+                    goto fail;
+                }
+            }
+            else {
+                LLVMValueRef maddr_base;
+
+                if (!(maddr_base =
+                          LLVMBuildIntToPtr(comp_ctx->builder, addr,
+                                            INT8_PTR_TYPE_GS, "maddr_base"))) {
+                    aot_set_last_error("llvm build int to ptr failed.");
+                    goto fail;
+                }
+                if (!(maddr2 = LLVMBuildInBoundsGEP2(
+                          comp_ctx->builder, INT8_TYPE, maddr_base,
+                          &offset_const, 1, "maddr"))) {
+                    aot_set_last_error("llvm build inboundgep failed.");
+                    goto fail;
+                }
+            }
+            LLVMBuildBr(comp_ctx->builder, merge_block);
+
+            SET_BUILD_POS(merge_block);
+            LLVMValueRef phi = LLVMBuildPhi(comp_ctx->builder, OPQ_PTR_TYPE, "phi");
+            LLVMValueRef incoming_values[] = {
+                maddr1, // Value from 'then' block
+                maddr2  // Value from 'else' block
+            };
+            LLVMBasicBlockRef incoming_blocks[] = { runtime_bounds_check, check_succ };
+            LLVMAddIncoming(phi, incoming_values, incoming_blocks, 2);
+
+            return phi;
```

Comment:
Don't use phi here? How about let aot_bounds_check return offset1 below?

Reply:
Because we need to check whether the offset is out of bound during runtime, a phi node is needed to select the offset dynamicly